### ISSUE TITLE
Support for dynamic still images and faster interface to provide data from an outside process.

### ISSUE THREAD

**Issue by pinballpower** [egnor/pivid#1]
Great project! I'm thinking about using it in a Pinball-related project as the code base seems to be a lot more stable than my own code.

However, there are a few things I would need:

1. Support for raw image data (RGBA). This is needed to render an image on top of the videos that gets updated at least 50 times per second.
2. A faster interface to send these data. I'm thinking about something like a named pipe or shared memory.

I'm happy to look into these and contribute code. However, I first want to understand if this is interesting for your project and make sure, it fits correctly into the overall architecture.

BTW: I'm happy to discuss the details off-line if you wish.

**Comment by egnor** [comment]
Intriguing! (I spent a while trying to figure out what software package "pinball" was until I realized that it's referring to _literal_ pinball, like with the little steel balls bouncing around a table, ha.) I expect for a lot of use cases there will be the desire to generate dynamic images from other processes and feed it in. The trick will be to avoid turning into another implementation of X windows (or Wayland) or generally complicating things a lot.

Do you have thoughts on what the best protocol would be? Ideally it's something that
- could easily be done from e.g. Python without auxiliary libraries
- but also passes memory directly, so zero-copy is possible
- can work in "immediate" mode (display whatever images are sent ASAP) and also "timed" mode (a sequence of frames with display timestamps for well-synchronized output)

The first two might seem impossible, but Python does support both mmap and unix-domain fd-passing natively. (Nothing's magical about Python here, it's just my proxy for "popular, approachable language", i.e. "not C++".) So it's theoretically possible!

So, maybe something like this?
- in the pivid play script, allow defining "live" or "external" media files
- instead of opening a static file, pivid creates & listens on a unix domain datagram socket in the media tree
- clients connect to that unix domain socket and send dma-buf file descriptors along with a short control messages
- the control message (JSON??) gives image pixel format (fourcc) & size & display timestamp (0 for "show asap")
- (clients can get dma-buf file descriptors from `/dev/dma_heap/linux,cma`)
- we include a little demo Python program to create & send a test animation (live clock?), as an example

Not having actually done it I'm not 100% sure this would work, but it seems like it might?

**Comment by pinballpower** [comment]
Yes, we're really talking about REAL Pinballs here ;-) Modern machines already have video screens with this kind of functionality, but the project linked above is about bringing this stuff to OLD machines.

While I did a lot of Python stuff in the past, I personally don't need a Python interface here (the existing stuff is C++). However, things like shared memory or named pipes are basically language agnostic and should therefore work with many programming languages. And I understand that sample code might be easier to understand in Python. I just guess, Python codes that does low-level DMA stuff might also not be very easy to understand for most people ;-)

I'm not sure how to exchange DMA descriptors, but I should be able to do this if there is some sample code around. I never did this kind of low-level stuff before. But hey, I also didn't do any C++ code for almost 30 years ;-)

Right now I'm still trying to figure out what the limits are that a Pi4 can handle. When adding more layers, at some point the stuff doesn't run properly anymore. The screen goes black for some time, the picture comes back again and so on. I had the same issue in my own code and couldn't figure out what to do about it and if there is a clear indication what are the limits in term of layers/videos/resolution/bandwidth. I saw some bandwidth calculations in your code, but I'm not sure if this can be used to determine the limits of the DRM. But this is something we probably should discuss offline.

**Comment by egnor** [comment]
(Started a new issue about managing/documenting hardware limits better: https://github.com/egnor/pivid/issues/3)

**Comment by egnor** [comment]
(Split off discussion of hardware limits: https://github.com/egnor/pivid/issues/3)
